src/join.rs

```diff
@@ -113,6 +129,98 @@ impl<J: Join> std::iter::Iterator for JoinIter<J> {
     }
 }
 
+/// `JoinParIter` is a `ParallelIterator` over a group of `Storages`.
+#[must_use]
+pub struct JoinParIter<J>(J);
+
+impl<J> ParallelIterator for JoinParIter<J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: Split + Send,
+    J::Mask: Send + Sync,
+{
+    type Item = J::Type;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+        where C: UnindexedConsumer<Self::Item>
+    {
+        let (keys, values) = self.0.open();
+        let producer = BitProducer((&keys).iter());
+        let values = UnsafeCell::new(values);
+        bridge_unindexed(JoinProducer::<J>::new(producer, &values), consumer)
+    }
+}
+
+struct JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    keys: BitProducer<'a, J::Mask>,
+    values: &'a UnsafeCell<J::Value>,
+}
+
+impl<'a, J> JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    fn new(keys: BitProducer<'a, J::Mask>, values: &'a UnsafeCell<J::Value>) -> Self {
+        JoinProducer {
+            keys: keys,
+            values: values,
+        }
+    }
+}
+
+unsafe impl<'a, J> Send for JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{}
+
+impl<'a, J> UnindexedProducer for JoinProducer<'a, J>
+  where
+    J: Join + Send,
+    J::Type: Send,
+    J::Value: 'a + Send,
+    J::Mask: 'a + Send + Sync,
+{
+    type Item = J::Type;
+    fn split(self) -> (Self, Option<Self>) {
+        let (cur, other) = self.keys.split();
+        let values = self.values;
+        let first = JoinProducer::new(cur, values);
+        let second = other.map(|o| JoinProducer::new(o, values));
+        (first, second)
+    }
+
+    fn fold_with<F>(self, folder: F) -> F
+      where
+        F: Folder<Self::Item>
+    {
+        let JoinProducer {values, keys, ..} = self;
+        let iter = keys.0.map(|idx| unsafe {
+            // This unsafe block should be safe if the `J::get`
+            // can be safely called from different threads with distinct indices.
+
+            // The `J::get` gets called for distinct indices because splitting happens by `BitProducer::split`
+            // which should ensure that all splitted parts are distinct.
+
+            // TODO: There can be multiple mutable references to storage and thus this is not safe, right?
```

Comment:
@WaDelma I'm aware of the wording in the documentation. Let's see what @kvark thinks about it. I don't see a reason why it shouldn't work.